Feature: multi-line logs

Currently when an multi-line event is written to a logfile promtail will take this as each row is its own entry and send the separately to loki. 

It would be great if Loki could handle multi-line events due to stacktraces.

This might also be a bug if promtail already should handle multi-lines.

Java stacktraces are always multiline. 

One of the problems with not having access to the full event is that it makes it much harder to search for other similar events because that the linebreaks can cut javaclasses in half and thus does not paint a real picture of what happened in the logs on the server.

Another thing is if the example below is handled with multiple logs instead it is really hard to search for say:   
- I know that these two keywords should be present somewhere in the stacktrace.  Say ```com.example.myproject.MyProjectServletException``` and ```javax.servlet.ServletException: Something bad happened```.

If I search for them withing having the stacktrace in one entry I will just get the rows that matched my keywords and not the whole stacktrace. This makes troubleshooting java much harder with Loki if multiline events are handled as-is. Because I first need to search for the keywords and then try to find the first and last entry of the stacktrace so I can read it as a whole. And the stacktraces can be several hundred lines.

I think there are other multiline events that should benefit from having it all in the same entry, but I have only worked with Java that behaves like that so I can not talk for other languages.


Here is an example stacktrace from java:
```
javax.servlet.ServletException: Something bad happened
    at com.example.myproject.OpenSessionInViewFilter.doFilter(OpenSessionInViewFilter.java:60)
    at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
    at com.example.myproject.ExceptionHandlerFilter.doFilter(ExceptionHandlerFilter.java:28)
    at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
    at com.example.myproject.OutputBufferFilter.doFilter(OutputBufferFilter.java:33)
    at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
    at org.mortbay.jetty.servlet.ServletHandler.handle(ServletHandler.java:388)
    at org.mortbay.jetty.security.SecurityHandler.handle(SecurityHandler.java:216)
    at org.mortbay.jetty.servlet.SessionHandler.handle(SessionHandler.java:182)
    at org.mortbay.jetty.handler.ContextHandler.handle(ContextHandler.java:765)
    at org.mortbay.jetty.webapp.WebAppContext.handle(WebAppContext.java:418)
    at org.mortbay.jetty.handler.HandlerWrapper.handle(HandlerWrapper.java:152)
    at org.mortbay.jetty.Server.handle(Server.java:326)
    at org.mortbay.jetty.HttpConnection.handleRequest(HttpConnection.java:542)
    at org.mortbay.jetty.HttpConnection$RequestHandler.content(HttpConnection.java:943)
    at org.mortbay.jetty.HttpParser.parseNext(HttpParser.java:756)
    at org.mortbay.jetty.HttpParser.parseAvailable(HttpParser.java:218)
    at org.mortbay.jetty.HttpConnection.handle(HttpConnection.java:404)
    at org.mortbay.jetty.bio.SocketConnector$Connection.run(SocketConnector.java:228)
    at org.mortbay.thread.QueuedThreadPool$PoolThread.run(QueuedThreadPool.java:582)
Caused by: com.example.myproject.MyProjectServletException
    at com.example.myproject.MyServlet.doPost(MyServlet.java:169)
    at javax.servlet.http.HttpServlet.service(HttpServlet.java:727)
    at javax.servlet.http.HttpServlet.service(HttpServlet.java:820)
    at org.mortbay.jetty.servlet.ServletHolder.handle(ServletHolder.java:511)
    at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1166)
    at com.example.myproject.OpenSessionInViewFilter.doFilter(OpenSessionInViewFilter.java:30)
    ... 27 more
Caused by: org.hibernate.exception.ConstraintViolationException: could not insert: [com.example.myproject.MyEntity]
    at org.hibernate.exception.SQLStateConverter.convert(SQLStateConverter.java:96)
    at org.hibernate.exception.JDBCExceptionHelper.convert(JDBCExceptionHelper.java:66)
    at org.hibernate.id.insert.AbstractSelectingDelegate.performInsert(AbstractSelectingDelegate.java:64)
    at org.hibernate.persister.entity.AbstractEntityPersister.insert(AbstractEntityPersister.java:2329)
    at org.hibernate.persister.entity.AbstractEntityPersister.insert(AbstractEntityPersister.java:2822)
    at org.hibernate.action.EntityIdentityInsertAction.execute(EntityIdentityInsertAction.java:71)
    at org.hibernate.engine.ActionQueue.execute(ActionQueue.java:268)
    at org.hibernate.event.def.AbstractSaveEventListener.performSaveOrReplicate(AbstractSaveEventListener.java:321)
    at org.hibernate.event.def.AbstractSaveEventListener.performSave(AbstractSaveEventListener.java:204)
    at org.hibernate.event.def.AbstractSaveEventListener.saveWithGeneratedId(AbstractSaveEventListener.java:130)
    at org.hibernate.event.def.DefaultSaveOrUpdateEventListener.saveWithGeneratedOrRequestedId(DefaultSaveOrUpdateEventListener.java:210)
    at org.hibernate.event.def.DefaultSaveEventListener.saveWithGeneratedOrRequestedId(DefaultSaveEventListener.java:56)
    at org.hibernate.event.def.DefaultSaveOrUpdateEventListener.entityIsTransient(DefaultSaveOrUpdateEventListener.java:195)
    at org.hibernate.event.def.DefaultSaveEventListener.performSaveOrUpdate(DefaultSaveEventListener.java:50)
    at org.hibernate.event.def.DefaultSaveOrUpdateEventListener.onSaveOrUpdate(DefaultSaveOrUpdateEventListener.java:93)
    at org.hibernate.impl.SessionImpl.fireSave(SessionImpl.java:705)
    at org.hibernate.impl.SessionImpl.save(SessionImpl.java:693)
    at org.hibernate.impl.SessionImpl.save(SessionImpl.java:689)
    at sun.reflect.GeneratedMethodAccessor5.invoke(Unknown Source)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:25)
    at java.lang.reflect.Method.invoke(Method.java:597)
    at org.hibernate.context.ThreadLocalSessionContext$TransactionProtectionWrapper.invoke(ThreadLocalSessionContext.java:344)
    at $Proxy19.save(Unknown Source)
    at com.example.myproject.MyEntityService.save(MyEntityService.java:59) <-- relevant call (see notes below)
    at com.example.myproject.MyServlet.doPost(MyServlet.java:164)
    ... 32 more
Caused by: java.sql.SQLException: Violation of unique constraint MY_ENTITY_UK_1: duplicate value(s) for column(s) MY_COLUMN in statement [...]
    at org.hsqldb.jdbc.Util.throwError(Unknown Source)
    at org.hsqldb.jdbc.jdbcPreparedStatement.executeUpdate(Unknown Source)
    at com.mchange.v2.c3p0.impl.NewProxyPreparedStatement.executeUpdate(NewProxyPreparedStatement.java:105)
    at org.hibernate.id.insert.AbstractSelectingDelegate.performInsert(AbstractSelectingDelegate.java:57)
    ... 54 more
```

Could you please encounter the following scenario or at least give your opinion:

Because of Loki we have changed the log format of our appenders to use the logstash JSON-layout-formatter in our logback configuration. That's why now we have an all information containing log entry with also the whole stacktrace (containing the already noted \n\t). So no need to join lines afterwards. That sounds good to me...

Here is a sample entry:

{"@timestamp":"2019-08-13T23:57:20.971+00:00","@version":"1","message":"Error in job timer thread occurred!","logger_name":"somecode.core.internal.job.Timer","thread_name":"JobTimer759282143","level":"ERROR","level_value":40000,"stack_trace":"somecode.core.capi.common.SystemException: java.sql.SQLException: Exception occurred while getting connection: oracle.ucp.UniversalConnectionPoolException: Cannot get Connection from Datasource: java.sql.SQLException: Listener refused the connection with the following error:\nORA-12505, TNS:listener does not currently know of SID given in connect descriptor\n \n\tat somecode.core.internal.jdbc.JDBCDataSourceMgrImpl.getConnection(JDBCDataSourceMgrImpl.java:783)\n\tat somecode.core.internal.jdbc.JDBCDataSourceMgrImpl.getConnection(JDBCDataSourceMgrImpl.java:796)\n\tat somecode.core.internal.jdbc.JDBCUtilsImpl.getConnection(JDBCUtilsImpl.java:925)\n\tat somecode.core.capi.jdbc.JDBCUtils.getConnection(JDBCUtils.java:560)\n\tat somecode.core.internal.jdbc.DBDate.currentDBDate(DBDate.java:75)\n\tat somecode.core.internal.job.Timer.run(Timer.java:102)\nCaused by: java.sql.SQLException: Exception occurred while getting connection: oracle.ucp.UniversalConnectionPoolException: Cannot get Connection from Datasource: java.sql.SQLException: Listener refused the connection with the following error:\nORA-12505, TNS:listener does not currently know of SID given in connect descriptor\n \n\tat oracle.ucp.util.UCPErrorHandler.newSQLException(UCPErrorHandler.java:479)\n\tat oracle.ucp.util.UCPErrorHandler.throwSQLException(UCPErrorHandler.java:154)\n\tat oracle.ucp.jdbc.PoolDataSourceImpl.getConnection(PoolDataSourceImpl.java:1127)\n\tat oracle.ucp.jdbc.PoolDataSourceImpl.getConnection(PoolDataSourceImpl.java:1031)\n\tat oracle.ucp.jdbc.PoolDataSourceImpl.getConnection(PoolDataSourceImpl.java:1006)\n\tat somecode.core.internal.jdbc.JDBCDataSource.getConnection(JDBCDataSource.java:101)\n\tat somecode.core.internal.jdbc.JDBCDataSource.getConnection(JDBCDataSource.java:73)\n\tat somecode.core.internal.jdbc.JDBCDataSourceMgrImpl.getConnection(JDBCDataSourceMgrImpl.java:779)\n\t... 5 common frames omitted\nCaused by: oracle.ucp.UniversalConnectionPoolException: Cannot get Connection from Datasource: java.sql.SQLException: Listener refused the connection with the following error:\nORA-12505, TNS:listener does not currently know of SID given in connect descriptor\n \n\tat oracle.ucp.util.UCPErrorHandler.newUniversalConnectionPoolException(UCPErrorHandler.java:359)\n\tat oracle.ucp.util.UCPErrorHandler.throwUniversalConnectionPoolException(UCPErrorHandler.java:58)\n\tat oracle.ucp.jdbc.DataSourceConnectionFactoryAdapter.createConnection(DataSourceConnectionFactoryAdapter.java:102)\n\tat oracle.ucp.jdbc.oracle.OracleDataSourceConnectionFactoryAdapter.createConnection(OracleDataSourceConnectionFactoryAdapter.java:82)\n\tat oracle.ucp.common.Database.createPooledConnection(Database.java:268)\n\tat oracle.ucp.common.Cluster.create(Cluster.java:234)\n\tat oracle.ucp.common.Core.growBorrowed(Core.java:634)\n\tat oracle.ucp.common.UniversalConnectionPoolImpl.borrowConnectionWithoutCountingRequests(UniversalConnectionPoolImpl.java:230)\n\tat oracle.ucp.common.UniversalConnectionPoolImpl.borrowConnectionAndValidate(UniversalConnectionPoolImpl.java:139)\n\tat oracle.ucp.common.UniversalConnectionPoolImpl.borrowConnection(UniversalConnectionPoolImpl.java:120)\n\tat oracle.ucp.jdbc.JDBCConnectionPool.borrowConnection(JDBCConnectionPool.java:170)\n\tat oracle.ucp.jdbc.oracle.OracleJDBCConnectionPool.borrowConnection(OracleJDBCConnectionPool.java:849)\n\tat oracle.ucp.jdbc.oracle.OracleConnectionConnectionPool.borrowConnection(OracleConnectionConnectionPool.java:82)\n\tat oracle.ucp.jdbc.PoolDataSourceImpl.getConnection(PoolDataSourceImpl.java:1103)\n\t... 10 common frames omitted\nCaused by: java.sql.SQLException: Listener refused the connection with the following error:\nORA-12505, TNS:listener does not currently know of SID given in connect descriptor\n \n\tat oracle.jdbc.driver.T4CConnection.logon(T4CConnection.java:743)\n\tat oracle.jdbc.driver.PhysicalConnection.connect(PhysicalConnection.java:666)\n\tat oracle.jdbc.driver.T4CDriverExtension.getConnection(T4CDriverExtension.java:32)\n\tat oracle.jdbc.driver.OracleDriver.connect(OracleDriver.java:566)\n\tat oracle.jdbc.pool.OracleDataSource.getPhysicalConnection(OracleDataSource.java:317)\n\tat oracle.jdbc.pool.OracleDataSource.getConnection(OracleDataSource.java:241)\n\tat oracle.ucp.jdbc.DataSourceConnectionFactoryAdapter.createConnection(DataSourceConnectionFactoryAdapter.java:96)\n\t... 21 common frames omitted\nCaused by: oracle.net.ns.NetException: Listener refused the connection with the following error:\nORA-12505, TNS:listener does not currently know of SID given in connect descriptor\n \n\tat oracle.net.ns.NSProtocolStream.negotiateConnection(NSProtocolStream.java:275)\n\tat oracle.net.ns.NSProtocol.connect(NSProtocol.java:264)\n\tat oracle.jdbc.driver.T4CConnection.connect(T4CConnection.java:1452)\n\tat oracle.jdbc.driver.T4CConnection.logon(T4CConnection.java:496)\n\t... 27 common frames omitted\n","executionsite":"root"}

But... In loki and the file log itself the stacktrace isn't readable anymore. That's not good. At least for the loki side maybe something like a configuration possiblity (fluentd plugin plus promtail) for fields would be good (e.g. multiline_fields: [stack_trace]). Then in grafana each multiline field could be represented in a well formated column maybe (just a simple thought).

I don't know if the one-json-line approach is the best one but just wanted to note it here. In theory it might also be well displayed.

PS: Also this issue https://github.com/grafana/loki/issues/616 also seems to be  related with multilines.

--- a/docs/sources/clients/promtail/pipelines.md
+++ b/docs/sources/clients/promtail/pipelines.md
@@ -206,6 +206,7 @@ Parsing stages:
 
 Transform stages:
 
+  - [multiline](../stages/multiline/): Merges multiple lines, e.g. stack traces, into multiline blocks.
   - [template](../stages/template/): Use Go templates to modify extracted data.
 
 Action stages:

--- a/docs/sources/clients/promtail/stages/multiline.md
+++ b/docs/sources/clients/promtail/stages/multiline.md
@@ -4,18 +4,63 @@ title: multiline
 
 # `multiline` stage
 
-...
+The `multiline` stage multiple lines into a multiline block before passing it on to the next stage in the pipeline.
+
+A new block is identified by the `firstline` regular expression. Any line that does *not* match the expression is considered to be part of the block of the previous match.
 
 ## Schema
 
 ```yaml
 multiline:
-  # TODO: document
+  # RE2 regular expression, if matched will start a new multiline block.
+  # This expresion must be provided.
   firstline: <string>
 
-  # TODO: document
-  max_lines: <integer>
-
-  # TODO: document
+  # The maximum wait time will be parsed as a Go duration: https://golang.org/pkg/time/#ParseDuration.
+  # If now new logs arrive withing this maximum wait time the current block will be sent on.
+  # This is useful if the opserved application dies with e.g. an exception. No new logs will arrive and the exception
+  # block is sent *after* the maximum wait time expired.
   max_wait_time: <duration>
+
+  # Maximum number of lines a block can have. If block has more lines a new block is started.
+  # The default is 128 lines.
+  max_lines: <integer>
+```
+
+## Examples
+
+Let's say we have the following logs from a very simple [flask](https://flask.palletsprojects.com) service.
+
+```
+[2020-12-03 11:36:20] "GET /hello HTTP/1.1" 200 -
+[2020-12-03 11:36:23] ERROR in app: Exception on /error [GET]
+Traceback (most recent call last):
+  File "/home/pallets/.pyenv/versions/3.8.5/lib/python3.8/site-packages/flask/app.py", line 2447, in wsgi_app
+    response = self.full_dispatch_request()
+  File "/home/pallets/.pyenv/versions/3.8.5/lib/python3.8/site-packages/flask/app.py", line 1952, in full_dispatch_request
+    rv = self.handle_user_exception(e)
+  File "/home/pallets/.pyenv/versions/3.8.5/lib/python3.8/site-packages/flask/app.py", line 1821, in handle_user_exception
+    reraise(exc_type, exc_value, tb)
+  File "/home/pallets/.pyenv/versions/3.8.5/lib/python3.8/site-packages/flask/_compat.py", line 39, in reraise
+    raise value
+  File "/home/pallets/.pyenv/versions/3.8.5/lib/python3.8/site-packages/flask/app.py", line 1950, in full_dispatch_request
+    rv = self.dispatch_request()
+  File "/home/pallets/.pyenv/versions/3.8.5/lib/python3.8/site-packages/flask/app.py", line 1936, in dispatch_request
+    return self.view_functions[rule.endpoint](**req.view_args)
+  File "/home/pallets/src/deployment_tools/hello.py", line 10, in error
+    raise Exception("Sorry, this route always breaks")
+Exception: Sorry, this route always breaks
+[2020-12-03 11:36:23] "GET /error HTTP/1.1" 500 -
+[2020-12-03 11:36:26] "GET /hello HTTP/1.1" 200 -
+[2020-12-03 11:36:27] "GET /hello HTTP/1.1" 200 -
+```
+
+We would like to collapse all lines of the traceback into one multiline block. All blocks start with a timestamp in brackets. Thus we configure a `multiline` stage with the `firstline` regular expression `^\[\d{4}-\d{2}-\d{2} \d{1,2}:\d{2}:\d{2}\]`. This will match the start of the traceback but not the following lines until `Exception: Sorry, this route always breaks`. These will be part of a multiline block and one log entry in Loki.
+
+```yaml
+multiline:
+  # Identify timestamps as first line of a multiline block.
+  firstline: "^\[\d{4}-\d{2}-\d{2} \d{1,2}:\d{2}:\d{2}\]"
+
+  max_wait_time: 3s
 ```

--- a/docs/sources/clients/promtail/stages/multiline.md
+++ b/docs/sources/clients/promtail/stages/multiline.md
@@ -20,6 +20,7 @@ multiline:
   # If now new logs arrive withing this maximum wait time the current block will be sent on.
   # This is useful if the opserved application dies with e.g. an exception. No new logs will arrive and the exception
   # block is sent *after* the maximum wait time expired.
+  # It defaults to 3s.
   max_wait_time: <duration>
 
   # Maximum number of lines a block can have. If block has more lines a new block is started.

--- a/docs/sources/clients/promtail/stages/multiline.md
+++ b/docs/sources/clients/promtail/stages/multiline.md
@@ -1,0 +1,21 @@
+---
+title: multiline 
+---
+
+# `multiline` stage
+
+...
+
+## Schema
+
+```yaml
+multiline:
+  # TODO: document
+  firstline: <string>
+
+  # TODO: document
+  max_lines: <integer>
+
+  # TODO: document
+  max_wait_time: <duration>
+```

--- a/pkg/logentry/stages/multiline.go
+++ b/pkg/logentry/stages/multiline.go
@@ -175,7 +175,7 @@ func (m *multilineStage) runMultiline(in chan Entry, out chan Entry, wg *sync.Wa
 	}
 }
 
-func (m *multilineStage) flush(out chan Entry, s *multilineState) {
+flush(out chan Entry, s *multilineState) {
 	if s.buffer.Len() == 0 {
 		level.Debug(m.logger).Log("msg", "nothing to flush", "buffer_len", s.buffer.Len())
 		return

--- a/pkg/logentry/stages/multiline.go
+++ b/pkg/logentry/stages/multiline.go
@@ -2,10 +2,14 @@ package stages
 
 import (
 	"bytes"
+	"fmt"
 	"regexp"
 	"time"
 
 	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/grafana/loki/pkg/logproto"
+	"github.com/grafana/loki/pkg/promtail/api"
 	"github.com/mitchellh/mapstructure"
 	"github.com/pkg/errors"
 )
@@ -25,8 +29,7 @@ type MultilineConfig struct {
 }
 
 func validateMultilineConfig(cfg *MultilineConfig) error {
-	if cfg == nil ||
-		(cfg.Expression == nil) {
+	if cfg == nil || cfg.Expression == nil || cfg.MaxWaitTime == nil {
 		return errors.New(ErrMultilineStageEmptyConfig)
 	}
 
@@ -79,9 +82,12 @@ func (m *multilineStage) Run(in chan Entry) chan Entry {
 		for {
 			select {
 			case <- time.After(m.cfg.maxWait):
+				level.Debug(m.logger).Log("msg", fmt.Sprintf("flush multiline block due to %v timeout", m.cfg.maxWait), "block", m.buffer.String())
 				m.flush(out)
 			case e, ok := <- in:
 				if !ok {
+					level.Debug(m.logger).Log("msg", "flush multiline block because inbound closed", "block", m.buffer.String())
+					m.flush(out)
 					return	
 				}
 
@@ -105,14 +111,19 @@ func (m *multilineStage) Run(in chan Entry) chan Entry {
 
 func (m *multilineStage) flush(out chan Entry) {
 	if m.buffer.Len() == 0 {
+		level.Debug(m.logger).Log("msg", "nothing to flush", "buffer_len", m.buffer.Len())
 		return
 	}
 
 	collapsed := &Entry{
-		Labels: m.startLineEntry.Labels,
 		Extracted: m.startLineEntry.Extracted,
-		Timestamp: m.startLineEntry.Timestamp,
-		Line: m.buffer.String(),
+		Entry: api.Entry{
+			Labels: m.startLineEntry.Entry.Labels,
+			Entry: logproto.Entry{
+				Timestamp: m.startLineEntry.Entry.Entry.Timestamp,
+				Line: m.buffer.String(),
+			},
+		},
 	}
 	m.buffer.Reset()
 

--- a/pkg/logentry/stages/multiline.go
+++ b/pkg/logentry/stages/multiline.go
@@ -102,6 +102,13 @@ func (m *multilineStage) Run(in chan Entry) chan Entry {
 			key := e.Labels.FastFingerprint()
 			s, ok := streams[key]
 			if !ok {
+				// Pass through entries until we hit first start line.
+				if !m.cfg.regex.MatchString(e.Line) {
+					level.Debug(m.logger).Log("msg", "pass through entry", "stream", key)
+					out <- e
+					continue
+				}
+
 				level.Debug(m.logger).Log("msg", "creating new stream", "stream", key)
 				s = make(chan Entry)
 				streams[key] = s

--- a/pkg/logentry/stages/multiline.go
+++ b/pkg/logentry/stages/multiline.go
@@ -23,7 +23,10 @@ const (
 	ErrMultilineStageInvalidMaxWaitTime = "multiline stage `max_wait_time` parse error: %v"
 )
 
-const maxLineDefault uint64 = 128
+const (
+	maxLineDefault uint64 = 128
+	maxWaitDefault = 3 * time.Second
+)
 
 // MultilineConfig contains the configuration for a multilineStage
 type MultilineConfig struct {
@@ -35,7 +38,7 @@ type MultilineConfig struct {
 }
 
 func validateMultilineConfig(cfg *MultilineConfig) error {
-	if cfg == nil || cfg.Expression == nil || cfg.MaxWaitTime == nil {
+	if cfg == nil || cfg.Expression == nil {
 		return errors.New(ErrMultilineStageEmptyConfig)
 	}
 
@@ -45,11 +48,15 @@ func validateMultilineConfig(cfg *MultilineConfig) error {
 	}
 	cfg.regex = expr
 
-	maxWait, err := time.ParseDuration(*cfg.MaxWaitTime)
-	if err != nil {
-		return errors.Errorf(ErrMultilineStageInvalidMaxWaitTime, err)
+	if cfg.MaxWaitTime != nil {
+		maxWait, err := time.ParseDuration(*cfg.MaxWaitTime)
+		if err != nil {
+			return errors.Errorf(ErrMultilineStageInvalidMaxWaitTime, err)
+		}
+		cfg.maxWait = maxWait
+	} else {
+		cfg.maxWait = maxWaitDefault
 	}
-	cfg.maxWait = maxWait
 
 	if cfg.MaxLines == nil {
 		cfg.MaxLines = new(uint64)
@@ -181,7 +188,7 @@ flush(out chan Entry, s *multilineState) {
 		return
 	}
 
-	collapsed := &Entry{
+	collapsed := Entry{
 		Extracted: s.startLineEntry.Extracted,
 		Entry: api.Entry{
 			Labels: s.startLineEntry.Entry.Labels,
@@ -194,7 +201,7 @@ flush(out chan Entry, s *multilineState) {
 	s.buffer.Reset()
 	s.currentLines = 0
 
-	out <- *collapsed
+	out <- collapsed
 }
 
 // Name implements Stage

--- a/pkg/logentry/stages/multiline.go
+++ b/pkg/logentry/stages/multiline.go
@@ -26,7 +26,7 @@ const maxLineDefault uint64 = 128
 type MultilineConfig struct {
 	Expression  *string `mapstructure:"firstline"`
 	regex       *regexp.Regexp
-	MaxLines     *uint64 `mapstructure:"max_lines"`
+	MaxLines    *uint64 `mapstructure:"max_lines"`
 	MaxWaitTime *string `mapstructure:"max_wait_time"`
 	maxWait     time.Duration
 }
@@ -58,11 +58,15 @@ func validateMultilineConfig(cfg *MultilineConfig) error {
 
 // dropMultiline matches lines to determine whether the following lines belong to a block and should be collapsed
 type multilineStage struct {
-	logger         log.Logger
-	cfg            *MultilineConfig
-	buffer         *bytes.Buffer
-	startLineEntry Entry
-	currentLines   uint64
+	logger log.Logger
+	cfg    *MultilineConfig
+}
+
+// multilineState captures the internal state of a running multiline stage.
+type multilineState struct {
+	buffer         *bytes.Buffer // The lines of the current multiline block.
+	startLineEntry Entry         // The entry of the start line of a multiline block.
+	currentLines   uint64        // The number of lines of the current multiline block.
 }
 
 // newMulitlineStage creates a MulitlineStage from config
@@ -80,7 +84,6 @@ func newMultilineStage(logger log.Logger, config interface{}) (Stage, error) {
 	return &multilineStage{
 		logger: log.With(logger, "component", "stage", "type", "multiline"),
 		cfg:    cfg,
-		buffer: new(bytes.Buffer),
 	}, nil
 }
 
@@ -88,34 +91,42 @@ func (m *multilineStage) Run(in chan Entry) chan Entry {
 	out := make(chan Entry)
 	go func() {
 		defer close(out)
+
+		state := &multilineState{
+			buffer:       new(bytes.Buffer),
+			currentLines: 0,
+		}
+
 		for {
 			select {
 			case <-time.After(m.cfg.maxWait):
-				level.Debug(m.logger).Log("msg", fmt.Sprintf("flush multiline block due to %v timeout", m.cfg.maxWait), "block", m.buffer.String())
-				m.flush(out)
+				level.Debug(m.logger).Log("msg", fmt.Sprintf("flush multiline block due to %v timeout", m.cfg.maxWait), "block", state.buffer.String())
+				m.flush(out, state)
 			case e, ok := <-in:
 				if !ok {
-					level.Debug(m.logger).Log("msg", "flush multiline block because inbound closed", "block", m.buffer.String())
-					m.flush(out)
+					level.Debug(m.logger).Log("msg", "flush multiline block because inbound closed", "block", state.buffer.String())
+					m.flush(out, state)
 					return
 				}
 
 				isFirstLine := m.cfg.regex.MatchString(e.Line)
 				if isFirstLine {
-					m.flush(out)
-					// TODO: we only consider the labels and timestamp from the firt entry. Should merge all entries?
-					m.startLineEntry = e
+					m.flush(out, state)
+
+					// The start line entry is used to set timestamp and labels in the flush method.
+					// The timestamps for following lines are ignored for now.
+					state.startLineEntry = e
 				}
 
 				// Append block line
-				if m.buffer.Len() > 0 {
-					m.buffer.WriteRune('\n')
+				if state.buffer.Len() > 0 {
+					state.buffer.WriteRune('\n')
 				}
-				m.buffer.WriteString(e.Line)
-				m.currentLines++
+				state.buffer.WriteString(e.Line)
+				state.currentLines++
 
-				if m.currentLines == *m.cfg.MaxLines {
-					m.flush(out)
+				if state.currentLines == *m.cfg.MaxLines {
+					m.flush(out, state)
 				}
 			}
 		}
@@ -123,24 +134,24 @@ func (m *multilineStage) Run(in chan Entry) chan Entry {
 	return out
 }
 
-func (m *multilineStage) flush(out chan Entry) {
-	if m.buffer.Len() == 0 {
-		level.Debug(m.logger).Log("msg", "nothing to flush", "buffer_len", m.buffer.Len())
+func (m *multilineStage) flush(out chan Entry, s *multilineState) {
+	if s.buffer.Len() == 0 {
+		level.Debug(m.logger).Log("msg", "nothing to flush", "buffer_len", s.buffer.Len())
 		return
 	}
 
 	collapsed := &Entry{
-		Extracted: m.startLineEntry.Extracted,
+		Extracted: s.startLineEntry.Extracted,
 		Entry: api.Entry{
-			Labels: m.startLineEntry.Entry.Labels,
+			Labels: s.startLineEntry.Entry.Labels,
 			Entry: logproto.Entry{
-				Timestamp: m.startLineEntry.Entry.Entry.Timestamp,
-				Line:      m.buffer.String(),
+				Timestamp: s.startLineEntry.Entry.Entry.Timestamp,
+				Line:      s.buffer.String(),
 			},
 		},
 	}
-	m.buffer.Reset()
-	m.currentLines = 0
+	s.buffer.Reset()
+	s.currentLines = 0
 
 	out <- *collapsed
 }

--- a/pkg/logentry/stages/multiline.go
+++ b/pkg/logentry/stages/multiline.go
@@ -4,6 +4,7 @@ import (
 	"bytes"
 	"fmt"
 	"regexp"
+	"sync"
 	"time"
 
 	"github.com/go-kit/kit/log"
@@ -12,6 +13,7 @@ import (
 	"github.com/grafana/loki/pkg/promtail/api"
 	"github.com/mitchellh/mapstructure"
 	"github.com/pkg/errors"
+	"github.com/prometheus/common/model"
 )
 
 const (
@@ -92,46 +94,77 @@ func (m *multilineStage) Run(in chan Entry) chan Entry {
 	go func() {
 		defer close(out)
 
-		state := &multilineState{
-			buffer:       new(bytes.Buffer),
-			currentLines: 0,
-		}
+		streams := make(map[model.Fingerprint](chan Entry))
+		wg := new(sync.WaitGroup)
 
-		for {
-			select {
-			case <-time.After(m.cfg.maxWait):
-				level.Debug(m.logger).Log("msg", fmt.Sprintf("flush multiline block due to %v timeout", m.cfg.maxWait), "block", state.buffer.String())
-				m.flush(out, state)
-			case e, ok := <-in:
-				if !ok {
-					level.Debug(m.logger).Log("msg", "flush multiline block because inbound closed", "block", state.buffer.String())
-					m.flush(out, state)
-					return
-				}
+		for e := range in {
+			key := e.Labels.FastFingerprint()
+			s, ok := streams[key]
+			if !ok {
+				level.Debug(m.logger).Log("msg", "creating new stream", "stream", key)
+				s = make(chan Entry)
+				streams[key] = s
 
-				isFirstLine := m.cfg.regex.MatchString(e.Line)
-				if isFirstLine {
-					m.flush(out, state)
-
-					// The start line entry is used to set timestamp and labels in the flush method.
-					// The timestamps for following lines are ignored for now.
-					state.startLineEntry = e
-				}
-
-				// Append block line
-				if state.buffer.Len() > 0 {
-					state.buffer.WriteRune('\n')
-				}
-				state.buffer.WriteString(e.Line)
-				state.currentLines++
-
-				if state.currentLines == *m.cfg.MaxLines {
-					m.flush(out, state)
-				}
+				wg.Add(1)
+				go m.runMultiline(s, out, wg)
 			}
+			level.Debug(m.logger).Log("msg", "pass entry", "stream", key, "line", e.Line)
+			s <- e
 		}
+
+		// Close all streams and wait for them to finish being processed.
+		for _, s := range streams {
+			close(s)
+		}
+		wg.Wait()
 	}()
 	return out
+}
+
+func (m *multilineStage) runMultiline(in chan Entry, out chan Entry, wg *sync.WaitGroup) {
+	defer wg.Done()
+
+	state := &multilineState{
+		buffer:       new(bytes.Buffer),
+		currentLines: 0,
+	}
+
+	for {
+		select {
+		case <-time.After(m.cfg.maxWait):
+			level.Debug(m.logger).Log("msg", fmt.Sprintf("flush multiline block due to %v timeout", m.cfg.maxWait), "block", state.buffer.String())
+			m.flush(out, state)
+		case e, ok := <-in:
+			level.Debug(m.logger).Log("msg", "processing line", "line", e.Line, "stream", e.Labels.FastFingerprint())
+
+			if !ok {
+				level.Debug(m.logger).Log("msg", "flush multiline block because inbound closed", "block", state.buffer.String(), "stream", e.Labels.FastFingerprint())
+				m.flush(out, state)
+				return
+			}
+
+			isFirstLine := m.cfg.regex.MatchString(e.Line)
+			if isFirstLine {
+				level.Debug(m.logger).Log("msg", "flush multiline block because new start line", "block", state.buffer.String(), "stream", e.Labels.FastFingerprint())
+				m.flush(out, state)
+
+				// The start line entry is used to set timestamp and labels in the flush method.
+				// The timestamps for following lines are ignored for now.
+				state.startLineEntry = e
+			}
+
+			// Append block line
+			if state.buffer.Len() > 0 {
+				state.buffer.WriteRune('\n')
+			}
+			state.buffer.WriteString(e.Line)
+			state.currentLines++
+
+			if state.currentLines == *m.cfg.MaxLines {
+				m.flush(out, state)
+			}
+		}
+	}
 }
 
 func (m *multilineStage) flush(out chan Entry, s *multilineState) {

--- a/pkg/logentry/stages/multiline.go
+++ b/pkg/logentry/stages/multiline.go
@@ -20,12 +20,15 @@ const (
 	ErrMultilineStageInvalidMaxWaitTime = "multiline stage `max_wait_time` parse error: %v"
 )
 
+const maxLineDefault uint64 = 128
+
 // MultilineConfig contains the configuration for a multilineStage
 type MultilineConfig struct {
 	Expression  *string `mapstructure:"firstline"`
+	regex       *regexp.Regexp
+	MaxLines     *uint64 `mapstructure:"max_lines"`
 	MaxWaitTime *string `mapstructure:"max_wait_time"`
 	maxWait     time.Duration
-	regex       *regexp.Regexp
 }
 
 func validateMultilineConfig(cfg *MultilineConfig) error {
@@ -45,6 +48,11 @@ func validateMultilineConfig(cfg *MultilineConfig) error {
 	}
 	cfg.maxWait = maxWait
 
+	if cfg.MaxLines == nil {
+		cfg.MaxLines = new(uint64)
+		*cfg.MaxLines = maxLineDefault
+	}
+
 	return nil
 }
 
@@ -54,6 +62,7 @@ type multilineStage struct {
 	cfg            *MultilineConfig
 	buffer         *bytes.Buffer
 	startLineEntry Entry
+	currentLines   uint64
 }
 
 // newMulitlineStage creates a MulitlineStage from config
@@ -103,6 +112,11 @@ func (m *multilineStage) Run(in chan Entry) chan Entry {
 					m.buffer.WriteRune('\n')
 				}
 				m.buffer.WriteString(e.Line)
+				m.currentLines++
+
+				if m.currentLines == *m.cfg.MaxLines {
+					m.flush(out)
+				}
 			}
 		}
 	}()
@@ -126,6 +140,7 @@ func (m *multilineStage) flush(out chan Entry) {
 		},
 	}
 	m.buffer.Reset()
+	m.currentLines = 0
 
 	out <- *collapsed
 }

--- a/pkg/logentry/stages/multiline.go
+++ b/pkg/logentry/stages/multiline.go
@@ -15,15 +15,15 @@ import (
 )
 
 const (
-	ErrMultilineStageEmptyConfig  = "multiline stage config must define `firstline` regular expression"
-	ErrMultilineStageInvalidRegex = "multiline stage first line regex compilation error: %v"
+	ErrMultilineStageEmptyConfig        = "multiline stage config must define `firstline` regular expression"
+	ErrMultilineStageInvalidRegex       = "multiline stage first line regex compilation error: %v"
 	ErrMultilineStageInvalidMaxWaitTime = "multiline stage `max_wait_time` parse error: %v"
 )
 
 // MultilineConfig contains the configuration for a multilineStage
 type MultilineConfig struct {
 	Expression  *string `mapstructure:"firstline"`
-	MaxWaitTime *string  `mapstructure:"max_wait_time"`
+	MaxWaitTime *string `mapstructure:"max_wait_time"`
 	maxWait     time.Duration
 	regex       *regexp.Regexp
 }
@@ -50,9 +50,9 @@ func validateMultilineConfig(cfg *MultilineConfig) error {
 
 // dropMultiline matches lines to determine whether the following lines belong to a block and should be collapsed
 type multilineStage struct {
-	logger log.Logger
-	cfg    *MultilineConfig
-	buffer *bytes.Buffer
+	logger         log.Logger
+	cfg            *MultilineConfig
+	buffer         *bytes.Buffer
 	startLineEntry Entry
 }
 
@@ -81,14 +81,14 @@ func (m *multilineStage) Run(in chan Entry) chan Entry {
 		defer close(out)
 		for {
 			select {
-			case <- time.After(m.cfg.maxWait):
+			case <-time.After(m.cfg.maxWait):
 				level.Debug(m.logger).Log("msg", fmt.Sprintf("flush multiline block due to %v timeout", m.cfg.maxWait), "block", m.buffer.String())
 				m.flush(out)
-			case e, ok := <- in:
+			case e, ok := <-in:
 				if !ok {
 					level.Debug(m.logger).Log("msg", "flush multiline block because inbound closed", "block", m.buffer.String())
 					m.flush(out)
-					return	
+					return
 				}
 
 				isFirstLine := m.cfg.regex.MatchString(e.Line)
@@ -121,7 +121,7 @@ func (m *multilineStage) flush(out chan Entry) {
 			Labels: m.startLineEntry.Entry.Labels,
 			Entry: logproto.Entry{
 				Timestamp: m.startLineEntry.Entry.Entry.Timestamp,
-				Line: m.buffer.String(),
+				Line:      m.buffer.String(),
 			},
 		},
 	}

--- a/pkg/logentry/stages/multiline.go
+++ b/pkg/logentry/stages/multiline.go
@@ -25,7 +25,7 @@ const (
 
 const (
 	maxLineDefault uint64 = 128
-	maxWaitDefault = 3 * time.Second
+	maxWaitDefault        = 3 * time.Second
 )
 
 // MultilineConfig contains the configuration for a multilineStage
@@ -196,7 +196,7 @@ func (m *multilineStage) runMultiline(in chan Entry, out chan Entry, wg *sync.Wa
 	}
 }
 
-func(m *multilineStage) flush(out chan Entry, s *multilineState) {
+func (m *multilineStage) flush(out chan Entry, s *multilineState) {
 	if s.buffer.Len() == 0 {
 		if Debug {
 			level.Debug(m.logger).Log("msg", "nothing to flush", "buffer_len", s.buffer.Len())

--- a/pkg/logentry/stages/multiline.go
+++ b/pkg/logentry/stages/multiline.go
@@ -1,0 +1,125 @@
+package stages
+
+import (
+	"bytes"
+	"regexp"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/mitchellh/mapstructure"
+	"github.com/pkg/errors"
+)
+
+const (
+	ErrMultilineStageEmptyConfig  = "multiline stage config must define `firstline` regular expression"
+	ErrMultilineStageInvalidRegex = "multiline stage first line regex compilation error: %v"
+	ErrMultilineStageInvalidMaxWaitTime = "multiline stage `max_wait_time` parse error: %v"
+)
+
+// MultilineConfig contains the configuration for a multilineStage
+type MultilineConfig struct {
+	Expression  *string `mapstructure:"firstline"`
+	MaxWaitTime *string  `mapstructure:"max_wait_time"`
+	maxWait     time.Duration
+	regex       *regexp.Regexp
+}
+
+func validateMultilineConfig(cfg *MultilineConfig) error {
+	if cfg == nil ||
+		(cfg.Expression == nil) {
+		return errors.New(ErrMultilineStageEmptyConfig)
+	}
+
+	expr, err := regexp.Compile(*cfg.Expression)
+	if err != nil {
+		return errors.Errorf(ErrMultilineStageInvalidRegex, err)
+	}
+	cfg.regex = expr
+
+	maxWait, err := time.ParseDuration(*cfg.MaxWaitTime)
+	if err != nil {
+		return errors.Errorf(ErrMultilineStageInvalidMaxWaitTime, err)
+	}
+	cfg.maxWait = maxWait
+
+	return nil
+}
+
+// dropMultiline matches lines to determine whether the following lines belong to a block and should be collapsed
+type multilineStage struct {
+	logger log.Logger
+	cfg    *MultilineConfig
+	buffer *bytes.Buffer
+	startLineEntry Entry
+}
+
+// newMulitlineStage creates a MulitlineStage from config
+func newMultilineStage(logger log.Logger, config interface{}) (Stage, error) {
+	cfg := &MultilineConfig{}
+	err := mapstructure.WeakDecode(config, cfg)
+	if err != nil {
+		return nil, err
+	}
+	err = validateMultilineConfig(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	return &multilineStage{
+		logger: log.With(logger, "component", "stage", "type", "multiline"),
+		cfg:    cfg,
+		buffer: new(bytes.Buffer),
+	}, nil
+}
+
+func (m *multilineStage) Run(in chan Entry) chan Entry {
+	out := make(chan Entry)
+	go func() {
+		defer close(out)
+		for {
+			select {
+			case <- time.After(m.cfg.maxWait):
+				m.flush(out)
+			case e, ok := <- in:
+				if !ok {
+					return	
+				}
+
+				isFirstLine := m.cfg.regex.MatchString(e.Line)
+				if isFirstLine {
+					m.flush(out)
+					// TODO: we only consider the labels and timestamp from the firt entry. Should merge all entries?
+					m.startLineEntry = e
+				}
+
+				// Append block line
+				if m.buffer.Len() > 0 {
+					m.buffer.WriteRune('\n')
+				}
+				m.buffer.WriteString(e.Line)
+			}
+		}
+	}()
+	return out
+}
+
+func (m *multilineStage) flush(out chan Entry) {
+	if m.buffer.Len() == 0 {
+		return
+	}
+
+	collapsed := &Entry{
+		Labels: m.startLineEntry.Labels,
+		Extracted: m.startLineEntry.Extracted,
+		Timestamp: m.startLineEntry.Timestamp,
+		Line: m.buffer.String(),
+	}
+	m.buffer.Reset()
+
+	out <- *collapsed
+}
+
+// Name implements Stage
+func (m *multilineStage) Name() string {
+	return StageTypeMultiline
+}

--- a/pkg/logentry/stages/multiline.go
+++ b/pkg/logentry/stages/multiline.go
@@ -111,19 +111,25 @@ func (m *multilineStage) Run(in chan Entry) chan Entry {
 			if !ok {
 				// Pass through entries until we hit first start line.
 				if !m.cfg.regex.MatchString(e.Line) {
-					level.Debug(m.logger).Log("msg", "pass through entry", "stream", key)
+					if Debug {
+						level.Debug(m.logger).Log("msg", "pass through entry", "stream", key)
+					}
 					out <- e
 					continue
 				}
 
-				level.Debug(m.logger).Log("msg", "creating new stream", "stream", key)
+				if Debug {
+					level.Debug(m.logger).Log("msg", "creating new stream", "stream", key)
+				}
 				s = make(chan Entry)
 				streams[key] = s
 
 				wg.Add(1)
 				go m.runMultiline(s, out, wg)
 			}
-			level.Debug(m.logger).Log("msg", "pass entry", "stream", key, "line", e.Line)
+			if Debug {
+				level.Debug(m.logger).Log("msg", "pass entry", "stream", key, "line", e.Line)
+			}
 			s <- e
 		}
 
@@ -147,20 +153,28 @@ func (m *multilineStage) runMultiline(in chan Entry, out chan Entry, wg *sync.Wa
 	for {
 		select {
 		case <-time.After(m.cfg.maxWait):
-			level.Debug(m.logger).Log("msg", fmt.Sprintf("flush multiline block due to %v timeout", m.cfg.maxWait), "block", state.buffer.String())
+			if Debug {
+				level.Debug(m.logger).Log("msg", fmt.Sprintf("flush multiline block due to %v timeout", m.cfg.maxWait), "block", state.buffer.String())
+			}
 			m.flush(out, state)
 		case e, ok := <-in:
-			level.Debug(m.logger).Log("msg", "processing line", "line", e.Line, "stream", e.Labels.FastFingerprint())
+			if Debug {
+				level.Debug(m.logger).Log("msg", "processing line", "line", e.Line, "stream", e.Labels.FastFingerprint())
+			}
 
 			if !ok {
-				level.Debug(m.logger).Log("msg", "flush multiline block because inbound closed", "block", state.buffer.String(), "stream", e.Labels.FastFingerprint())
+				if Debug {
+					level.Debug(m.logger).Log("msg", "flush multiline block because inbound closed", "block", state.buffer.String(), "stream", e.Labels.FastFingerprint())
+				}
 				m.flush(out, state)
 				return
 			}
 
 			isFirstLine := m.cfg.regex.MatchString(e.Line)
 			if isFirstLine {
-				level.Debug(m.logger).Log("msg", "flush multiline block because new start line", "block", state.buffer.String(), "stream", e.Labels.FastFingerprint())
+				if Debug {
+					level.Debug(m.logger).Log("msg", "flush multiline block because new start line", "block", state.buffer.String(), "stream", e.Labels.FastFingerprint())
+				}
 				m.flush(out, state)
 
 				// The start line entry is used to set timestamp and labels in the flush method.
@@ -182,9 +196,11 @@ func (m *multilineStage) runMultiline(in chan Entry, out chan Entry, wg *sync.Wa
 	}
 }
 
-func flush(out chan Entry, s *multilineState) {
+func(m *multilineStage) flush(out chan Entry, s *multilineState) {
 	if s.buffer.Len() == 0 {
-		level.Debug(m.logger).Log("msg", "nothing to flush", "buffer_len", s.buffer.Len())
+		if Debug {
+			level.Debug(m.logger).Log("msg", "nothing to flush", "buffer_len", s.buffer.Len())
+		}
 		return
 	}
 

--- a/pkg/logentry/stages/multiline.go
+++ b/pkg/logentry/stages/multiline.go
@@ -182,7 +182,7 @@ func (m *multilineStage) runMultiline(in chan Entry, out chan Entry, wg *sync.Wa
 	}
 }
 
-flush(out chan Entry, s *multilineState) {
+func flush(out chan Entry, s *multilineState) {
 	if s.buffer.Len() == 0 {
 		level.Debug(m.logger).Log("msg", "nothing to flush", "buffer_len", s.buffer.Len())
 		return

--- a/pkg/logentry/stages/multiline.go
+++ b/pkg/logentry/stages/multiline.go
@@ -9,11 +9,12 @@ import (
 
 	"github.com/go-kit/kit/log"
 	"github.com/go-kit/kit/log/level"
-	"github.com/grafana/loki/pkg/logproto"
-	"github.com/grafana/loki/pkg/promtail/api"
 	"github.com/mitchellh/mapstructure"
 	"github.com/pkg/errors"
 	"github.com/prometheus/common/model"
+
+	"github.com/grafana/loki/pkg/logproto"
+	"github.com/grafana/loki/pkg/promtail/api"
 )
 
 const (

--- a/pkg/logentry/stages/multiline_test.go
+++ b/pkg/logentry/stages/multiline_test.go
@@ -33,15 +33,19 @@ func Test_multilineStage_Process(t *testing.T) {
 	}
 
 	out := processEntries(stage,
+		simpleEntry("not a start line before 1", "label"),
+		simpleEntry("not a start line before 2", "label"),
 		simpleEntry("START line 1", "label"),
 		simpleEntry("not a start line", "label"),
 		simpleEntry("START line 2", "label"),
 		simpleEntry("START line 3", "label"))
 
-	require.Len(t, out, 3)
-	require.Equal(t, "START line 1\nnot a start line", out[0].Line)
-	require.Equal(t, "START line 2", out[1].Line)
-	require.Equal(t, "START line 3", out[2].Line)
+	require.Len(t, out, 5)
+	require.Equal(t, "not a start line before 1", out[0].Line)
+	require.Equal(t, "not a start line before 2", out[1].Line)
+	require.Equal(t, "START line 1\nnot a start line", out[2].Line)
+	require.Equal(t, "START line 2", out[3].Line)
+	require.Equal(t, "START line 3", out[4].Line)
 }
 func Test_multilineStage_MultiStreams(t *testing.T) {
 	// Enable debug logging

--- a/pkg/logentry/stages/multiline_test.go
+++ b/pkg/logentry/stages/multiline_test.go
@@ -1,0 +1,47 @@
+package stages
+
+import (
+	"bytes"
+	"testing"
+	"time"
+
+	"github.com/cortexproject/cortex/pkg/util"
+	"github.com/prometheus/common/model"
+	"github.com/stretchr/testify/require"
+	ww "github.com/weaveworks/common/server"
+)
+
+func Test_multilineStage_Process(t *testing.T) {
+
+	// Enable debug logging
+	cfg := &ww.Config{}
+	require.Nil(t, cfg.LogLevel.Set("debug"))
+	util.InitLogger(cfg)
+	Debug = true
+
+	mcfg := &MultilineConfig{Expression: ptrFromString("^START")}
+	err := validateMultilineConfig(mcfg)
+	require.NoError(t, err)
+
+	stage := &multilineStage{
+		cfg:    mcfg,
+		logger: util.Logger,
+		buffer: new(bytes.Buffer),
+	}
+
+	out := processEntries(stage, simpleEntry("START line 1"), simpleEntry("not a start line"), simpleEntry("START line 2"), simpleEntry("START line 3"))
+
+	require.Equal(t, "START line 1\nnot a start line", out[0].Line)
+	require.Equal(t, "START line 2", out[1].Line)
+	require.Equal(t, "START line 3", out[2].Line)
+}
+
+func simpleEntry(line string) Entry {
+	return Entry{
+				Labels:    model.LabelSet{},
+				Line:      ptrFromString(line),
+				Extracted: map[string]interface{}{},
+				Timestamp: ptrFromTime(time.Now()),
+			}
+
+}

--- a/pkg/logentry/stages/multiline_test.go
+++ b/pkg/logentry/stages/multiline_test.go
@@ -1,7 +1,6 @@
 package stages
 
 import (
-	"bytes"
 	"sync"
 	"testing"
 	"time"
@@ -29,7 +28,6 @@ func Test_multilineStage_Process(t *testing.T) {
 	stage := &multilineStage{
 		cfg:    mcfg,
 		logger: util.Logger,
-		buffer: new(bytes.Buffer),
 	}
 
 	out := processEntries(stage, simpleEntry("START line 1"), simpleEntry("not a start line"), simpleEntry("START line 2"), simpleEntry("START line 3"))
@@ -54,7 +52,6 @@ func Test_multilineStage_MaxWaitTime(t *testing.T) {
 	stage := &multilineStage{
 		cfg:    mcfg,
 		logger: util.Logger,
-		buffer: new(bytes.Buffer),
 	}
 
 	in := make(chan Entry, 2)

--- a/pkg/logentry/stages/multiline_test.go
+++ b/pkg/logentry/stages/multiline_test.go
@@ -87,21 +87,21 @@ func Test_multilineStage_MaxWaitTime(t *testing.T) {
 		return
 	}()
 
-	require.Eventually(t, func() bool {mu.Lock(); defer mu.Unlock(); return len(res) == 2;}, time.Duration(3 * maxWait), time.Second)
+	require.Eventually(t, func() bool { mu.Lock(); defer mu.Unlock(); return len(res) == 2 }, time.Duration(3*maxWait), time.Second)
 	require.Equal(t, "START line", res[0].Line)
 	require.Equal(t, "not a start line hitting timeout", res[1].Line)
 }
 
 func simpleEntry(line string) Entry {
 	return Entry{
-				Extracted: map[string]interface{}{},
-				Entry: api.Entry{
-					Labels:    model.LabelSet{},
-					Entry: logproto.Entry{
-						Timestamp: time.Now(),
-						Line: line,
-					},
-				},
-			}
+		Extracted: map[string]interface{}{},
+		Entry: api.Entry{
+			Labels: model.LabelSet{},
+			Entry: logproto.Entry{
+				Timestamp: time.Now(),
+				Line:      line,
+			},
+		},
+	}
 
 }

--- a/pkg/logentry/stages/multiline_test.go
+++ b/pkg/logentry/stages/multiline_test.go
@@ -7,11 +7,12 @@ import (
 	"time"
 
 	"github.com/cortexproject/cortex/pkg/util"
-	"github.com/grafana/loki/pkg/logproto"
-	"github.com/grafana/loki/pkg/promtail/api"
 	"github.com/prometheus/common/model"
 	"github.com/stretchr/testify/require"
 	ww "github.com/weaveworks/common/server"
+
+	"github.com/grafana/loki/pkg/logproto"
+	"github.com/grafana/loki/pkg/promtail/api"
 )
 
 func Test_multilineStage_Process(t *testing.T) {
@@ -94,7 +95,7 @@ func Test_multilineStage_MaxWaitTime(t *testing.T) {
 	util.InitLogger(cfg)
 	Debug = true
 
-	maxWait := time.Duration(2 * time.Second)
+	maxWait := 2 * time.Second
 	mcfg := &MultilineConfig{Expression: ptrFromString("^START"), MaxWaitTime: ptrFromString(maxWait.String())}
 	err := validateMultilineConfig(mcfg)
 	require.NoError(t, err)
@@ -117,7 +118,6 @@ func Test_multilineStage_MaxWaitTime(t *testing.T) {
 			res = append(res, e)
 			mu.Unlock()
 		}
-		return
 	}()
 
 	// Write input with a delay
@@ -131,10 +131,9 @@ func Test_multilineStage_MaxWaitTime(t *testing.T) {
 
 		// Signal pipeline we are done.
 		close(in)
-		return
 	}()
 
-	require.Eventually(t, func() bool { mu.Lock(); defer mu.Unlock(); return len(res) == 2 }, time.Duration(3*maxWait), time.Second)
+	require.Eventually(t, func() bool { mu.Lock(); defer mu.Unlock(); return len(res) == 2 }, 3*maxWait, time.Second)
 	require.Equal(t, "START line", res[0].Line)
 	require.Equal(t, "not a start line hitting timeout", res[1].Line)
 }

--- a/pkg/logentry/stages/multiline_test.go
+++ b/pkg/logentry/stages/multiline_test.go
@@ -1,6 +1,7 @@
 package stages
 
 import (
+	"sort"
 	"sync"
 	"testing"
 	"time"
@@ -30,13 +31,62 @@ func Test_multilineStage_Process(t *testing.T) {
 		logger: util.Logger,
 	}
 
-	out := processEntries(stage, simpleEntry("START line 1"), simpleEntry("not a start line"), simpleEntry("START line 2"), simpleEntry("START line 3"))
+	out := processEntries(stage,
+		simpleEntry("START line 1", "label"),
+		simpleEntry("not a start line", "label"),
+		simpleEntry("START line 2", "label"),
+		simpleEntry("START line 3", "label"))
 
 	require.Len(t, out, 3)
 	require.Equal(t, "START line 1\nnot a start line", out[0].Line)
 	require.Equal(t, "START line 2", out[1].Line)
 	require.Equal(t, "START line 3", out[2].Line)
 }
+func Test_multilineStage_MultiStreams(t *testing.T) {
+	// Enable debug logging
+	cfg := &ww.Config{}
+	require.Nil(t, cfg.LogLevel.Set("debug"))
+	util.InitLogger(cfg)
+	Debug = true
+
+	mcfg := &MultilineConfig{Expression: ptrFromString("^START"), MaxWaitTime: ptrFromString("3s")}
+	err := validateMultilineConfig(mcfg)
+	require.NoError(t, err)
+
+	stage := &multilineStage{
+		cfg:    mcfg,
+		logger: util.Logger,
+	}
+
+	out := processEntries(stage,
+		simpleEntry("START line 1", "one"),
+		simpleEntry("not a start line 1", "one"),
+		simpleEntry("START line 1", "two"),
+		simpleEntry("not a start line 2", "one"),
+		simpleEntry("START line 2", "two"),
+		simpleEntry("START line 2", "one"),
+		simpleEntry("not a start line 1", "one"),
+	)
+
+	sort.Slice(out, func(l, r int) bool {
+		return out[l].Timestamp.Before(out[r].Timestamp)
+	})
+
+	require.Len(t, out, 4)
+
+	require.Equal(t, "START line 1\nnot a start line 1\nnot a start line 2", out[0].Line)
+	require.Equal(t, model.LabelValue("one"), out[0].Labels["value"])
+
+	require.Equal(t, "START line 1", out[1].Line)
+	require.Equal(t, model.LabelValue("two"), out[1].Labels["value"])
+
+	require.Equal(t, "START line 2", out[2].Line)
+	require.Equal(t, model.LabelValue("two"), out[2].Labels["value"])
+
+	require.Equal(t, "START line 2\nnot a start line 1", out[3].Line)
+	require.Equal(t, model.LabelValue("one"), out[3].Labels["value"])
+}
+
 func Test_multilineStage_MaxWaitTime(t *testing.T) {
 	// Enable debug logging
 	cfg := &ww.Config{}
@@ -72,12 +122,12 @@ func Test_multilineStage_MaxWaitTime(t *testing.T) {
 
 	// Write input with a delay
 	go func() {
-		in <- simpleEntry("START line")
+		in <- simpleEntry("START line", "label")
 
 		// Trigger flush due to max wait timeout
 		time.Sleep(2 * maxWait)
 
-		in <- simpleEntry("not a start line hitting timeout")
+		in <- simpleEntry("not a start line hitting timeout", "label")
 
 		// Signal pipeline we are done.
 		close(in)
@@ -89,11 +139,11 @@ func Test_multilineStage_MaxWaitTime(t *testing.T) {
 	require.Equal(t, "not a start line hitting timeout", res[1].Line)
 }
 
-func simpleEntry(line string) Entry {
+func simpleEntry(line, label string) Entry {
 	return Entry{
 		Extracted: map[string]interface{}{},
 		Entry: api.Entry{
-			Labels: model.LabelSet{},
+			Labels: model.LabelSet{"value": model.LabelValue(label)},
 			Entry: logproto.Entry{
 				Timestamp: time.Now(),
 				Line:      line,

--- a/pkg/logentry/stages/multiline_test.go
+++ b/pkg/logentry/stages/multiline_test.go
@@ -2,10 +2,13 @@ package stages
 
 import (
 	"bytes"
+	"sync"
 	"testing"
 	"time"
 
 	"github.com/cortexproject/cortex/pkg/util"
+	"github.com/grafana/loki/pkg/logproto"
+	"github.com/grafana/loki/pkg/promtail/api"
 	"github.com/prometheus/common/model"
 	"github.com/stretchr/testify/require"
 	ww "github.com/weaveworks/common/server"
@@ -19,7 +22,7 @@ func Test_multilineStage_Process(t *testing.T) {
 	util.InitLogger(cfg)
 	Debug = true
 
-	mcfg := &MultilineConfig{Expression: ptrFromString("^START")}
+	mcfg := &MultilineConfig{Expression: ptrFromString("^START"), MaxWaitTime: ptrFromString("3s")}
 	err := validateMultilineConfig(mcfg)
 	require.NoError(t, err)
 
@@ -31,17 +34,74 @@ func Test_multilineStage_Process(t *testing.T) {
 
 	out := processEntries(stage, simpleEntry("START line 1"), simpleEntry("not a start line"), simpleEntry("START line 2"), simpleEntry("START line 3"))
 
+	require.Len(t, out, 3)
 	require.Equal(t, "START line 1\nnot a start line", out[0].Line)
 	require.Equal(t, "START line 2", out[1].Line)
 	require.Equal(t, "START line 3", out[2].Line)
 }
+func Test_multilineStage_MaxWaitTime(t *testing.T) {
+	// Enable debug logging
+	cfg := &ww.Config{}
+	require.Nil(t, cfg.LogLevel.Set("debug"))
+	util.InitLogger(cfg)
+	Debug = true
+
+	maxWait := time.Duration(2 * time.Second)
+	mcfg := &MultilineConfig{Expression: ptrFromString("^START"), MaxWaitTime: ptrFromString(maxWait.String())}
+	err := validateMultilineConfig(mcfg)
+	require.NoError(t, err)
+
+	stage := &multilineStage{
+		cfg:    mcfg,
+		logger: util.Logger,
+		buffer: new(bytes.Buffer),
+	}
+
+	in := make(chan Entry, 2)
+	out := stage.Run(in)
+
+	// Accumulate result
+	mu := new(sync.Mutex)
+	var res []Entry
+	go func() {
+		for e := range out {
+			mu.Lock()
+			t.Logf("appending %s", e.Line)
+			res = append(res, e)
+			mu.Unlock()
+		}
+		return
+	}()
+
+	// Write input with a delay
+	go func() {
+		in <- simpleEntry("START line")
+
+		// Trigger flush due to max wait timeout
+		time.Sleep(2 * maxWait)
+
+		in <- simpleEntry("not a start line hitting timeout")
+
+		// Signal pipeline we are done.
+		close(in)
+		return
+	}()
+
+	require.Eventually(t, func() bool {mu.Lock(); defer mu.Unlock(); return len(res) == 2;}, time.Duration(3 * maxWait), time.Second)
+	require.Equal(t, "START line", res[0].Line)
+	require.Equal(t, "not a start line hitting timeout", res[1].Line)
+}
 
 func simpleEntry(line string) Entry {
 	return Entry{
-				Labels:    model.LabelSet{},
-				Line:      ptrFromString(line),
 				Extracted: map[string]interface{}{},
-				Timestamp: ptrFromTime(time.Now()),
+				Entry: api.Entry{
+					Labels:    model.LabelSet{},
+					Entry: logproto.Entry{
+						Timestamp: time.Now(),
+						Line: line,
+					},
+				},
 			}
 
 }

--- a/pkg/logentry/stages/stage.go
+++ b/pkg/logentry/stages/stage.go
@@ -27,6 +27,7 @@ const (
 	StageTypePipeline  = "pipeline"
 	StageTypeTenant    = "tenant"
 	StageTypeDrop      = "drop"
+	StageTypeMultiline = "multiline"
 )
 
 // Processor takes an existing set of labels, timestamp and log entry and returns either a possibly mutated
@@ -136,6 +137,11 @@ func New(logger log.Logger, jobName *string, stageType string,
 		}
 	case StageTypeDrop:
 		s, err = newDropStage(logger, cfg, registerer)
+		if err != nil {
+			return nil, err
+		}
+	case StageTypeMultiline:
+		s, err = newMultilineStage(logger, cfg)
 		if err != nil {
 			return nil, err
 		}

